pandasai/__init__.py

```diff
@@ -299,8 +299,15 @@ def conversational_answer(self, question: str, answer: str) -> str:
         return self._llm.call(generate_response_instruction, "")
 
     def _get_prompt(
-        self, key: str, default_prompt: Prompt, default_values: Dict = {}, df=None
-    ):
+        self,
+        key: str,
+        default_prompt: Type[Prompt],
+        default_values: Optional[dict] = None,
+        df=None,
+    ) -> tuple[Prompt, dict]:
+        if default_values is None:
+            default_values = {}
```

Comment:
@gventuri 
> Why not making `default_values: Optional[dict] = {}` instead? Also, I think `df` is actually unused. Shouldn't we rather remove it?

It'd be better to avoid dict and any mutual object as a default values for functions/methods. Kind of anti-pattern. The reason for this is the object is created only during the first call, then the same object is going to be used:
```python 
>>> def foobar(mutual_param={}, **kwargs):
...     mutual_param.update(**kwargs)
...     print(mutual_param)
>>>    
>>> foobar(a=1, b=2)
{'a': 1, 'b': 2}
>>> foobar(c=3, d=4)
{'a': 1, 'b': 2, 'c': 3, 'd': 4}  # the dictionary was not created again, instead the old one was chosen;
                                  # that potentially can lead to some bugs
                                  # when re-using PandasAI object with different prompts
```
Also see:
https://docs.quantifiedcode.com/python-anti-patterns/correctness/mutable_default_value_as_argument.html#anti-pattern
https://stackoverflow.com/questions/1132941/least-astonishment-and-the-mutable-default-argument
https://web.archive.org/web/20200221224620id_/http://effbot.org/zone/default-values.htm

___

>Also, I think df is actually unused. Shouldn't we rather remove it?

Same thought I! But lots of Ci tests were failed (you can see [here](https://github.com/gventuri/pandas-ai/actions/runs/5684010480/job/15405794934)), so i decided just to restore the `df` parameter and let it be as it was. IDK, maybe some tests had been passing `df` param to the method as far as i get from logs. Let me know if it rather should be investigated and fixed.

Reply:
Ok got it. If I try to remove it as an argument, we still have a test failing. That test is the one that tests that if we use $df.something in the prompt, it will then have access to the df var from the "global" scope. Now I made it so it only have access to selected variables (for now only df) so df is no more unused!